File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-riskdef-5-4/FormulaTerm.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class FormulaTerm : ISerialized
    {
        public FormulaTerm(XmlNode xmlNode)
        {
            XmlNode coefficientNode = xmlNode.SelectSingleNode("coefficient");
            
            if (coefficientNode != null)
            {
                if (coefficientNode.Attributes["href"] != null || coefficientNode.Attributes["id"] != null) 
                {
                    if (coefficientNode.Attributes["id"] != null) 
                    {
                        coefficientIDRef_ = coefficientNode.Attributes["id"].Value;
                        XsdTypeDecimal ob = new XsdTypeDecimal(coefficientNode);
                        IDManager.SetID(coefficientIDRef_, ob);
                    }
                    else if (coefficientNode.Attributes["href"] != null)
                    {
                        coefficientIDRef_ = coefficientNode.Attributes["href"].Value;
                    }
                    else
                    {
                        coefficient_ = new XsdTypeDecimal(coefficientNode);
                    }
                }
                else
                {
                    coefficient_ = new XsdTypeDecimal(coefficientNode);
                }
            }
            
        
            XmlNodeList partialDerivativeReferenceNodeList = xmlNode.SelectNodes("partialDerivativeReference");
            
            if (partialDerivativeReferenceNodeList != null)
            {
                this.partialDerivativeReference_ = new List<PricingParameterDerivativeReference>();
                foreach (XmlNode item in partialDerivativeReferenceNodeList)
                {
                    if (item.Attributes["href"] != null || item.Attributes["id"] != null) 
                    {
                        if (item.Attributes["id"] != null) 
                        {
                            partialDerivativeReferenceIDRef_ = item.Attributes["id"].Value;
                            partialDerivativeReference_.Add(new PricingParameterDerivativeReference(item));
                            IDManager.SetID(partialDerivativeReferenceIDRef_, partialDerivativeReference_[partialDerivativeReference_.Count - 1 ]);
                        }
                        else if (item.Attributes["href"] != null)
                        {
                            partialDerivativeReferenceIDRef_ = item.Attributes["href"].Value;
                        }
                        else
                        {
                        partialDerivativeReference_.Add(new PricingParameterDerivativeReference(item));
                        }
                    }
                    else
                    {
                        partialDerivativeReference_.Add(new PricingParameterDerivativeReference(item));
                    }
                }
            }
            
        
        }
        
    
        #region Coefficient_
        private XsdTypeDecimal coefficient_;
        public XsdTypeDecimal Coefficient_
        {
            get
            {
                if (this.coefficient_ != null)
                {
                    return this.coefficient_; 
                }
                else if (this.coefficientIDRef_ != null)
                {
                    coefficient_ = IDManager.getID(coefficientIDRef_) as XsdTypeDecimal;
                    return this.coefficient_; 
                }
                else
                {
                      return this.coefficient_; 
                }
            }
            set
            {
                if (this.coefficient_ != value)
                {
                    this.coefficient_ = value;
                }
            }
        }
        #endregion
        
        public string coefficientIDRef_ { get; set; }
        #region PartialDerivativeReference_
        private List<PricingParameterDerivativeReference> partialDerivativeReference_;
        public List<PricingParameterDerivativeReference> PartialDerivativeReference_
        {
            get
            {
                if (this.partialDerivativeReference_ != null)
                {
                    return this.partialDerivativeReference_; 
                }
                else if (this.partialDerivativeReferenceIDRef_ != null)
                {
                    return this.partialDerivativeReference_; 
                }
                else
                {
                      return this.partialDerivativeReference_; 
                }
            }
            set
            {
                if (this.partialDerivativeReference_ != value)
                {
                    this.partialDerivativeReference_ = value;
                }
            }
        }
        #endregion
        
        public string partialDerivativeReferenceIDRef_ { get; set; }
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-riskdef-5-4/FormulaTerm.cpp
// FormulaTerm.cpp 
#include "FormulaTerm.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

FormulaTerm::FormulaTerm(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //coefficientNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* coefficientNode = xmlNode->FirstChildElement("coefficient");

   if(coefficientNode){coefficientIsNull_ = false;}
   else{coefficientIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- coefficientNode , address : " << coefficientNode << std::endl;
   #endif
   if(coefficientNode)
   {
      if(coefficientNode->Attribute("href") || coefficientNode->Attribute("id"))
      {
          if(coefficientNode->Attribute("id"))
          {
             coefficientIDRef_ = coefficientNode->Attribute("id");
             coefficient_ = boost::shared_ptr<XsdTypeDecimal>(new XsdTypeDecimal(coefficientNode));
             coefficient_->setID(coefficientIDRef_);
             IDManager::instance().setID(coefficientIDRef_,coefficient_);
          }
          else if(coefficientNode->Attribute("href")) { coefficientIDRef_ = coefficientNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { coefficient_ = boost::shared_ptr<XsdTypeDecimal>(new XsdTypeDecimal(coefficientNode));}
   }

   //partialDerivativeReferenceNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* partialDerivativeReferenceNode = xmlNode->FirstChildElement("partialDerivativeReference");

   if(partialDerivativeReferenceNode){partialDerivativeReferenceIsNull_ = false;}
   else{partialDerivativeReferenceIsNull_ = true;}

   if(partialDerivativeReferenceNode)
   {
      for(partialDerivativeReferenceNode; partialDerivativeReferenceNode; partialDerivativeReferenceNode = partialDerivativeReferenceNode->NextSiblingElement("partialDerivativeReference")){
          #ifdef ConsolePrint
              FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- partialDerivativeReferenceNode , address : " << partialDerivativeReferenceNode << std::endl;
          #endif
          if(partialDerivativeReferenceNode->Attribute("href") || partialDerivativeReferenceNode->Attribute("id"))
          {
              if(partialDerivativeReferenceNode->Attribute("id"))
              {
                  partialDerivativeReferenceIDRef_ = partialDerivativeReferenceNode->Attribute("id");
                  partialDerivativeReference_.push_back(boost::shared_ptr<PricingParameterDerivativeReference>(new PricingParameterDerivativeReference(partialDerivativeReferenceNode)));
                  partialDerivativeReference_.back()->setID(partialDerivativeReferenceIDRef_);
                  IDManager::instance().setID(partialDerivativeReferenceIDRef_, partialDerivativeReference_.back());
              }
              else if(partialDerivativeReferenceNode->Attribute("href")) { partialDerivativeReferenceIDRef_ = partialDerivativeReferenceNode->Attribute("href");}
              else { QL_FAIL("id or href error"); }
          }
          else { partialDerivativeReference_.push_back(boost::shared_ptr<PricingParameterDerivativeReference>(new PricingParameterDerivativeReference(partialDerivativeReferenceNode)));}
      }
   }
   else {
       #ifdef ConsolePrint
           FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- partialDerivativeReferenceNode , address : " << partialDerivativeReferenceNode << std::endl;
       #endif
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<XsdTypeDecimal> FormulaTerm::getCoefficient()
{
   if(!this->coefficientIsNull_){
        if(coefficientIDRef_ != ""){
             return boost::shared_static_cast<XsdTypeDecimal>(IDManager::instance().getID(coefficientIDRef_));
        }else{
             return this->coefficient_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<XsdTypeDecimal>();
   }
}
std::vector<boost::shared_ptr<PricingParameterDerivativeReference>> FormulaTerm::getPartialDerivativeReference()
{
   if(!this->partialDerivativeReferenceIsNull_){
        if(partialDerivativeReferenceIDRef_ != ""){
             return this->partialDerivativeReference_;
        }else{
             return this->partialDerivativeReference_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return std::vector<boost::shared_ptr<PricingParameterDerivativeReference>>();
   }
}
}

